fem/bilininteg_divergence.cpp

```diff
@@ -0,0 +1,1073 @@
+                  QQQ2[qz][qy][qx] = w;
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(qz,z,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qy,y,Q1D)
+            {
+               MFEM_FOREACH_THREAD(qx,x,Q1D)
+               {
+                  const int q = qx + (qy + qz * Q1D) * Q1D;
+                  const double gX = QQQ0[qz][qy][qx];
+                  const double gY = QQQ1[qz][qy][qx];
+                  const double gZ = QQQ2[qz][qy][qx];
+                  div[qz][qy][qx] += gX*Q(q,0,c,e) + gY*Q(q,1,c,e) + gZ*Q(q,2,c,e);
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+
+      if (tidz == 0)
+      {
+         MFEM_FOREACH_THREAD(d,y,D1DE)
+         {
+            MFEM_FOREACH_THREAD(q,x,Q1D)
+            {
+               Bt[d][q] = bt(d,q);
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+
+      MFEM_FOREACH_THREAD(qz,z,Q1D)
+      {
+         MFEM_FOREACH_THREAD(qy,y,Q1D)
+         {
+            MFEM_FOREACH_THREAD(dx,x,D1DE)
+            {
+               double u = 0.0;
+               for (int qx = 0; qx < Q1D; ++qx)
+               {
+                  u += div[qz][qy][qx] * Bt[dx][qx];
+               }
+               QQD0[qz][qy][dx] = u;
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      MFEM_FOREACH_THREAD(qz,z,Q1D)
+      {
+         MFEM_FOREACH_THREAD(dy,y,D1DE)
+         {
+            MFEM_FOREACH_THREAD(dx,x,D1DE)
+            {
+               double u = 0.0;
+               for (int qy = 0; qy < Q1D; ++qy)
+               {
+                  u += QQD0[qz][qy][dx] * Bt[dy][qy];
+               }
+               QDD0[qz][dy][dx] = u;
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      MFEM_FOREACH_THREAD(dz,z,D1DE)
+      {
+         MFEM_FOREACH_THREAD(dy,y,D1DE)
+         {
+            MFEM_FOREACH_THREAD(dx,x,D1DE)
+            {
+               double u = 0.0;
+               for (int qz = 0; qz < Q1D; ++qz)
+               {
+                  u += QDD0[qz][dy][dx] * Bt[dz][qz];
+               }
+               y(dx,dy,dz,e) += u;
+            }
+         }
+      }
+   });
+}
+
+static void PADivergenceApply(const int dim,
+                              const int TR_D1D,
+                              const int TE_D1D,
+                              const int Q1D,
+                              const int NE,
+                              const Array<double> &B,
+                              const Array<double> &G,
+                              const Array<double> &Bt,
+                              const Vector &op,
+                              const Vector &x,
+                              Vector &y,
+                              bool transpose=false)
+{
+   if (dim == 2)
+   {
+      return PADivergenceApply2D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+   }
+   if (dim == 3)
+   {
+      return PADivergenceApply3D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+   }
```

Comment:
@tzanio @v-dobrev @camierjs We need to figure out a way of doing template specialization for common cases for these kernels without hugely increasing compile times. Here we're only using the default fallback kernel because otherwise (now that we're adding more and more kernels) the compilation is just too long. Also the `switch ((D1D << 4 ) | Q1D)` approach is error-prone and requires too much manual intervention (e.g. picking exactly the right Q1D to match the quadrature rule).